pkg/compact/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package garbagecollector
+
+import (
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/compact/downsample"
+)
+
+type GarbageBlocksFinder struct{}
+
+func NewGarbageBlocksFinder() *GarbageBlocksFinder {
+	return &GarbageBlocksFinder{}
+}
+
+func (s *GarbageBlocksFinder) GarbageBlocks(resolution int64, blocks map[ulid.ULID]*metadata.Meta) (ids []ulid.ULID, err error) {
+	// Map each block to its highest priority parent. Initial blocks have themselves
+	// in their source section, i.e. are their own parent.
+	parents := map[ulid.ULID]ulid.ULID{}
+
+	for id, meta := range blocks {
+		// Skip any block that has a different resolution.
+		if meta.Thanos.Downsample.Resolution != resolution {
+			continue
+		}
+
+		// For each source block we contain, check whether we are the highest priority parent block.
+		for _, sid := range meta.Compaction.Sources {
```

Comment:
This is confusing for me. How this is really different to our Duplicate block filter? :thinking: Isn't this exactly the same?